Add CPU_kernel with float16 for adjust_contrast_op.cc

SuryanarayanaY opened the pull request:
The API `tf.image.adjust_contrast` supports `float16` and `float32` for the argument images . But with `CPU` runtime and `float16` dtype the API raising exception stating there is no kernel available. Refer attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/6990939226d9d79dfe9225b13919c288/61246_cpu.ipynb) for the exception.

Hence adding a kernel for CPU with float16.

Shall also fixes #61246

SuryanarayanaY committed:
Add CPU_kernel with float16 for adjust_contrast_op.cc

The API tf.image.adjust_contrast supports float16 and float32 for the argument images . But with CPU runtime and float16 dtype the API raising exception stating there is no kernel available.

Hence adding a kernel for CPU with float16.

cantonios commented on tensorflow/core/kernels/image/adjust_contrast_op.cc:
```diff
@@ -387,9 +387,14 @@ class AdjustContrastOpv2<CPUDevice, float> : public AdjustContrastOpV2Base {
+      AdjustContrastOpv2<CPUDevice, T>);
+
+REGISTER_KERNEL(float);
+REGISTER_KERNEL(Eigen::half);
```
You are getting undefined templates from this:
```
tensorflow/core/kernels/image/adjust_contrast_op.cc:396:1: error: implicit instantiation of undefined template 'tensorflow::AdjustContrastOpv2<Eigen::ThreadPoolDevice, Eigen::half>'
  396 | REGISTER_KERNEL(Eigen::half);
      | ^
tensorflow/core/kernels/image/adjust_contrast_op.cc:393:7: note: expanded from macro 'REGISTER_KERNEL'
  393 |       AdjustContrastOpv2<CPUDevice, T>);
      |       ^
tensorflow/core/kernels/image/adjust_contrast_op.cc:202:7: note: template is declared here
  202 | class AdjustContrastOpv2;
      |       ^
1 error generated.
```

cantonios commented on tensorflow/core/kernels/image/adjust_contrast_op.cc:
```diff
@@ -387,9 +387,14 @@ class AdjustContrastOpv2<CPUDevice, float> : public AdjustContrastOpV2Base {
+#define REGISTER_KERNEL(T)                                                 \
+  REGISTER_KERNEL_BUILDER(                                                 \
+      Name("AdjustContrastv2").Device(DEVICE_CPU).TypeConstraint<T>("T"),  \
+      AdjustContrastOpv2<CPUDevice, T>);
```
nit: the trailing semicolon here leads to double semicolons when you call
```
REGISTER_KERNEL(...);
```
below.  You should remove it from here.

SuryanarayanaY replied:
Changes done as suggested to avoid double semicolons.

SuryanarayanaY committed:
Update adjust_contrast_op.cc

Removed ";"  at end of macro definition to avoid double semicolons since while calling I used ";" already.

SuryanarayanaY commented:
@cantonios ,

Thanks for the inputs. I got a doubt here. Suppose consider below snippet from identity_op.cc.

https://github.com/tensorflow/tensorflow/blob/9d2607201ed37fef0b3d23249a2f757a9047a680/tensorflow/core/kernels/identity_op.cc#L154-L158

In the above Op code we are using `;` at both Marco end and at Register_Kernel call as well. I am not sure how its not having problem for this case. Any pointers would be appreciated.

SuryanarayanaY commented:
@cantonios , Done the changes as suggested. Could you please review!

cantonios commented:
> @cantonios ,
> 
> Thanks for the inputs. I got a doubt here. Suppose consider below snippet from identity_op.cc.
> 
> https://github.com/tensorflow/tensorflow/blob/9d2607201ed37fef0b3d23249a2f757a9047a680/tensorflow/core/kernels/identity_op.cc#L154-L158
> 
> In the above Op code we are using `;` at both Marco end and at Register_Kernel call as well. I am not sure how its not having problem for this case. Any pointers would be appreciated.

It does have a problem in that case.  We are likely getting build warnings because of it.

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

SuryanarayanaY committed:
Syntax errors correction in adjust_contrast_op.cc

Syntax errors corrected as suggested

SuryanarayanaY commented:
Done the changes.

cantonios commented:
It looks like you still haven't resolved this issue:
```
implicit instantiation of undefined template 'tensorflow::AdjustContrastOpv2
```

gbaned commented:
Hi @SuryanarayanaY Can you please check @cantonios's comments ? Thank you!

gbaned commented:
Hi @SuryanarayanaY  Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

SuryanarayanaY commented:
Not stale.Looking into it.

SuryanarayanaY committed:
Implement AdjustContrastV2 OP for generic datatype T

Implemented AdjustContrastV2 OP for generic datatype T.

I left the factor to float itself as i think it is referring to contrast_factor which is registered for float only.

SuryanarayanaY commented:
Updated the kernel.

cantonios commented:
Can you add a test?

gbaned commented:
Hi @SuryanarayanaY Can you please check @cantonios's [comment](https://github.com/tensorflow/tensorflow/pull/61782#issuecomment-1990693272)? Thank you!
